pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,378 @@
+	if entry == nil {
+		return false, fmt.Errorf("process cache entry not found")
+	}
+	if !entry.HasCompleteLineage() && !shadowInsertion { // check that the process context lineage is complete, otherwise drop it
+		at.Stats.droppedCount[event.GetEventType()][brokenLineageReason].Inc()
+		return false, fmt.Errorf("incomplete lineage")
+	}
+
+	node, newProcessNode, err := at.CreateProcessNode(entry, Runtime, shadowInsertion)
+	if newProcessNode && shadowInsertion {
+		return true, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	if node == nil {
+		// a process node couldn't be found or created for this event, ignore it
+		return false, err
+	}
+
+	// resolve fields
+	event.ResolveFieldsForAD()
+
+	// the count of processed events is the count of events that matched the activity dump selector = the events for
+	// which we successfully found a process activity node
+	at.Stats.processedCount[event.GetEventType()].Inc()
+
+	// insert the event based on its type
+	switch event.GetEventType() {
+	case model.ExecEventType:
+		// tag the matched rules if any
+		node.MatchedRules = model.AppendMatchedRule(node.MatchedRules, event.Rules)
+		return newProcessNode, nil
+	case model.FileOpenEventType:
+		return node.InsertFileEvent(&event.Open.File, event, Runtime, at.Stats, at.shouldMergePaths, shadowInsertion)
+	case model.DNSEventType:
+		return node.InsertDNSEvent(event, at.Stats, at.DNSNames, shadowInsertion)
+	case model.BindEventType:
+		return node.InsertBindEvent(event, at.Stats, shadowInsertion)
+	case model.SyscallsEventType:
+		return node.InsertSyscalls(event, at.SyscallsMask)
+	}
+
+	return false, nil
+}
+
+// CreateProcessNode finds or a create a new process activity node in the activity dump if the entry
+// matches the activity dump selector.
+func (at *ActivityTree) CreateProcessNode(entry *model.ProcessCacheEntry, generationType NodeGenerationType, shadowInsertion bool) (node *ProcessNode, newProcessNode bool, err error) {
+	if entry == nil {
+		return nil, false, nil
+	}
+
+	// drop processes with abnormal paths
+	if entry.GetPathResolutionError() != "" {
+		return nil, false, fmt.Errorf("process cache entry error: %s", entry.GetPathResolutionError())
+	}
+
+	// look for a ProcessActivityNode by process cookie
+	if entry.Cookie > 0 {
+		var found bool
+		node, found = at.CookieToProcessNode[entry.Cookie]
+		if found {
+			return node, false, nil
+		}
+	}
+
+	defer func() {
+		// if a node was found, and if the entry has a valid cookie, insert a cookie shortcut
+		if entry.Cookie > 0 && node != nil {
+			at.CookieToProcessNode[entry.Cookie] = node
+		}
+	}()
+
+	// find or create a ProcessActivityNode for the parent of the input ProcessCacheEntry. If the parent is a fork entry,
+	// jump immediately to the next ancestor.
+	parentNode, newProcessNode, err := at.CreateProcessNode(entry.GetNextAncestorBinary(), Snapshot, shadowInsertion)
+	if err != nil || newProcessNode && shadowInsertion {
```

Comment:
just : `if err != nil || shadowInsertion` ?

Reply:
I think the `newProcessNode` check is important: unless I'm missing something, we want to check if we're modifying something in the graph. This includes making sure that the entire process tree is in the profile. We can return early as soon as (and only if) we see something new.